forte/data/extractor/predictor.py

```diff
@@ -0,0 +1,199 @@
+                example_collection[tag].append(feature)
+
+        tensor_collection = {}
+        for tag, features in example_collection.items():
+            converter = self.feature_scheme[tag]["converter"]
+            tensor, mask = converter.convert(features)
+
+            if tag not in tensor_collection:
+                tensor_collection[tag] = {}
+
+            tensor_collection[tag]["tensor"] = tensor
+            tensor_collection[tag]["mask"] = mask
+        return tensor_collection
+
+    def yield_batch(self, pack):
+        for instance in pack.get(self.scope):
+            feature_collection = {}
+            for tag, scheme in self.feature_scheme.items():
+                extractor = scheme["extractor"]
+                feature = extractor.extract(pack, instance)
+                feature_collection[tag] = feature
+            
+            self.pack_pools.append(pack)
+            self.instance_pools.append(instance)
+            self.features_pools.append(feature_collection)
+
+            if len(self.features_pools) == self.batch_size:
+                yield self.convert(self.features_pools[:self.batch_size]), \
+                    self.pack_pools[:self.batch_size], \
+                    self.instance_pools[:self.batch_size]
+                
+                self.features_pools.clear()
+                self.pack_pools.clear()
+                self.instance_pools.clear()
+
+    def flush_batch(self):
+        if len(self.features_pools) > 0:
+            yield self.convert(self.features_pools), self.pack_pools, \
+                self.instance_pools
+            self.features_pools.clear()
+            self.pack_pools.clear()
+            self.instance_pools.clear()
+
+    def _process(self, input_pack: PackType):
+        for tensor_collection, packs, instances in self.yield_batch(input_pack):
+            predictions = self.predict(tensor_collection)
+            self.add_to_pack(predictions, tensor_collection, packs, instances)
+
+        # update the status of the jobs. The jobs which were removed from
+        # data_pack_pool will have status "PROCESSED" else they are "QUEUED"
+        q_index = self._process_manager.current_queue_index
+        u_index = self._process_manager.unprocessed_queue_indices[q_index]
+        data_pool_length = len(set(self.pack_pools))
+        current_queue = self._process_manager.current_queue
+
+        for i, job_i in enumerate(
+                itertools.islice(current_queue, 0, u_index + 1)):
+            if i <= u_index - data_pool_length:
+                job_i.set_status(ProcessJobStatus.PROCESSED)
+            else:
+                job_i.set_status(ProcessJobStatus.QUEUED)
+
+    def flush(self):
+        for tensor_collection, packs, instances in self.flush_batch():
+            predictions = self.predict(tensor_collection)
+            self.add_to_pack(predictions, tensor_collection, packs, instances)
+            pack = packs[0]
```

Comment:
What does this line do?

Reply:
There is a pool to retain features that we get from the pack. Once the pool is filled with a batch of features, it will generate them. So the flush function is to generator all those features finally remained in the pool. So that they are not missed. I learn this from the original batcher class.

Reply:
but how does `pack = packs[0]` help?

Reply:
Sorry. It is useless. This is a line of code I didn't clean up.